src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Debug/ProjectDebuggerProvider.cs

```diff
@@ -0,0 +1,339 @@
+                    hStdError = unchecked((uint)info.StandardErrorHandle.ToInt32()),
+                    flags = (uint)__DSI_FLAGS.DSI_USESTDHANDLES,
+                };
+                debugInfo.pStartupInfo = Marshal.AllocCoTaskMem(Marshal.SizeOf(processStartupInfo));
+                Marshal.StructureToPtr(processStartupInfo, debugInfo.pStartupInfo, false);
+            }
+
+            debugInfo.AppPackageLaunchInfo = info.AppPackageLaunchInfo;
+            debugInfo.project = info.Project;
+
+            return debugInfo;
+        }
+
+        /// <summary>
+        /// Frees memory allocated by GetDebuggerStruct.
+        /// </summary>
+        internal static void FreeVsDebugTargetInfoStruct(VsDebugTargetInfo4 nativeStruct)
+        {
+            if (nativeStruct.pDebugEngines != IntPtr.Zero)
+            {
+                Marshal.FreeCoTaskMem(nativeStruct.pDebugEngines);
+            }
+
+            if (nativeStruct.pStartupInfo != IntPtr.Zero)
+            {
+                Marshal.DestroyStructure(nativeStruct.pStartupInfo, typeof(VsDebugStartupInfo));
+                Marshal.FreeCoTaskMem(nativeStruct.pStartupInfo);
+            }
+        }
+
+        /// <summary>
+        /// Converts the environment key value pairs to a valid environment string of the form
+        /// key=value/0key2=value2/0/0, with null's between each entry and a double null terminator.
+        /// </summary>
+        private static string GetSerializedEnvironmentString(IDictionary<string, string> environment)
+        {
+            // If no dictionary was set, or its empty, the debugger wants null for its environment block.
+            if (environment == null || environment.Count == 0)
+            {
+                return null;
+            }
+
+            // Collect all the variables as a null delimited list of key=value pairs.
+            StringBuilder result = new StringBuilder();
+            foreach (var pair in environment)
+            {
+                result.Append(pair.Key);
+                result.Append('=');
+                result.Append(pair.Value);
+                result.Append('\0');
+            }
+
+            // Add a final list-terminating null character.
+            // This is sent to native code as a BSTR and no null is added automatically.
+            // But the contract of the format of the data requires that this be a null-delimited,
+            // null-terminated list.
+            result.Append('\0');
+            return result.ToString();
+        }
+
+        /// <summary>
+        /// Collects an array of GUIDs into an array of bytes.
+        /// </summary>
+        /// <remarks>
+        /// The order of the GUIDs are preserved, and each GUID is copied exactly one after the other in the byte array.
+        /// </remarks>
+        private static byte[] GetGuidBytes(IList<Guid> guids)
+        {
+            int sizeOfGuid = Guid.Empty.ToByteArray().Length;
```

Comment:
I assume this is not a high frequency code so these allocations should not affect performance